infobip-typescript-generator-extension-api/src/main/java/com/infobip/typescript/transformer/ClassTransformerDecoratorExtension.java

```diff
@@ -59,29 +57,43 @@ private TsBeanModel decorateClass(SymbolTable symbolTable, TsBeanModel bean) {
     }
 
     private TsPropertyModel getDecorators(SymbolTable symbolTable, TsBeanModel model, TsPropertyModel tsPropertyModel) {
-        return getField(model, tsPropertyModel).map(Field::getType)
-                                               .filter(type -> Modifier.isAbstract(
-                                                       type.getModifiers()) || type.isInterface())
-                                               .map(type -> tsPropertyModel.withDecorators(
-                                                       getDecorators(symbolTable, model, tsPropertyModel, type)))
+        return getField(model, tsPropertyModel).map(this::getParameterizedTypeClasses)
+                                               .map(parameterizedTypeClasses -> tsPropertyModel.withDecorators(
+                                                       getDecorators(symbolTable, model, tsPropertyModel,
+                                                                     resolveTypeToDecorate(parameterizedTypeClasses))))
                                                .orElse(tsPropertyModel);
     }
 
+    private Class<?> resolveTypeToDecorate(ParameterizedTypeClasses parameterizedTypeClasses) {
+        Class<?> type = parameterizedTypeClasses.getType();
+        Class<?> typeToDecorate = type;
+        if (type.isArray()) {
+            typeToDecorate = type.getComponentType();
+        }
+
+        final Optional<Class<?>> typeArgument = parameterizedTypeClasses.getTypeArgument();
+        if (Collection.class.isAssignableFrom(type) && typeArgument.isPresent()) {
+            typeToDecorate = typeArgument.get();
+        }
+        return typeToDecorate;
+    }
+
     private List<TsDecorator> getDecorators(SymbolTable symbolTable,
                                             TsBeanModel model,
                                             TsPropertyModel tsPropertyModel,
                                             Class<?> type) {
         return factory.create(type)
                       .filter(resolver -> resolver instanceof CompositeJsonTypeResolver<?>)
+                      .filter(resolver -> !Modifier.isFinal(type.getModifiers()))
```

Comment:
why we have added this filter ?

Reply:
As you see below, these filters determine whether there will be a hierarchical `@Type` decorator like
```
@Type(() => Object, {
        discriminator: {
            property: 'type', subTypes: [
                { value: ReplySuggestion, name: SuggestionType.REPLY },
                { value: OpenUrlSuggestion, name: SuggestionType.OPEN_URL }
            ]
        }
    })
```
or a 'flat' one like this: `@Type(() => CardContent)`

I had a case where I had an array of objects of a type that extended `SimpleJsonHierarchy` so the above filter `resolver instanceof CompositeJsonTypeResolver<?>` returned true, but the type itself was final (it was a hierarchy 'leaf') so I just wanted the 'flat' decorator.

I'll do some refactoring here to improve readability.

Reply:
It's refactored now and hopefully my intensions are now more clear :)

Reply:
double check if this  !Modifier.isFinal(type.getModifiers()) is good enough and if it covers all cases